File: C++/pseudo_palindromic_paths_in_a_binary_tree.cpp
// no. 1457
#include "tree_node.h"
#include <vector>
using namespace std;

class Solution {
public:
  vector<int> value_count;
  int result;
  void traverse(TreeNode *root) {
    value_count[root->val - 1]++;
    if (!root->left && !root->right) {
      int num_odd = 0;
      for (int i : value_count) {
        if (i % 2) {
          num_odd++;
        }
      }
      if (num_odd <= 1) {
        result++;
      }
      value_count[root->val - 1]--;
      return;
    }
    if (root->left) {
      traverse(root->left);
    }
    if (root->right) {
      traverse(root->right);
    }
    value_count[root->val - 1]--;
  }
  int pseudoPalindromicPaths(TreeNode *root) {
    for (int i = 0; i < 9; i++) {
      value_count.push_back(0);
    }
    traverse(root);
    return result;
  }
};

File: python/pseudo_palindromic_paths_in_a_binary_tree.py
#no. 1457
from utils import TreeNode
class Solution:
    def __init__(self):
        self.value_count = [0 for _ in range(9)]
        self.result = 0

    def traverse(self, root):
        self.value_count[root.val - 1] += 1
        if not root.left and not root.right:
            #    print(self.value_count)
            num_odd = 0
            for i in self.value_count:
                if i % 2:
                    num_odd += 1
            if num_odd <= 1:
                self.result += 1
            self.value_count[root.val - 1] -= 1
            return
        if root.left:
            self.traverse(root.left)
        if root.right:
            self.traverse(root.right)
        self.value_count[root.val - 1] -= 1

    def pseudoPalindromicPaths(self, root: Optional[TreeNode]) -> int:
        self.traverse(root)
        return self.result

